source/rsa.c

```diff
@@ -286,3 +310,103 @@ int is_valid_rsa_key_size(size_t key_size_in_bits) {
 
     return AWS_OP_SUCCESS;
 }
+
+#ifndef BYO_CRYPTO
+
+static uint8_t s_rsa_encryption_oid[] = {0x2a, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x01, 0x01};
+
+/**
+ * There are likely tens of algo specifiers for rsa out there.
+ * But in practice mostly everyone uses rsaEncryption oid for key at rest.
+ * So for now just support that and we can add other ones later if needed.
+ * Even though its technically encryption oid, most crypto libs are lax
+ * and allow the key to be used for all operations.
+ */
+static struct aws_byte_cursor s_rsa_encryption_oid_cur = {
+    .ptr = (s_rsa_encryption_oid),
+    .len = sizeof(s_rsa_encryption_oid),
+};
+
+/**
+ * Win and Mac dont provide native support for loading pkcs8 keys, so
+ * for simplicity and consistency just parse pkcs1 key from pkcs8 and
+ * use it with existing pkcs1 loader.
+ */
+struct aws_rsa_key_pair *aws_rsa_key_pair_new_from_private_key_pkcs8_impl(
+    struct aws_allocator *allocator,
+    struct aws_byte_cursor private_key) {
+
+    struct aws_der_decoder *decoder = aws_der_decoder_new(allocator, private_key);
+
+    if (decoder == NULL) {
+        return NULL;
+    }
+
+    /**
+     * Format of pkcs8 is as follows.
+     * PrivateKeyInfo ::= SEQUENCE {
+     *  version         Version,          -- INTEGER (0)
+     *  algorithm       AlgorithmIdentifier,
+     *  privateKey      OCTET STRING,     -- contains DER encoded key
+     *  attributes  [0] IMPLICIT SET OF Attribute OPTIONAL
+     * }
+     * AlgorithmIdentifier ::= SEQUENCE {
+     *  algorithm       OBJECT IDENTIFIER,
+     *  parameters      ANY DEFINED BY algorithm OPTIONAL
+     * }
+     */
+
+    if (!aws_der_decoder_next(decoder) || aws_der_decoder_tlv_type(decoder) != AWS_DER_SEQUENCE) {
+        aws_raise_error(AWS_ERROR_CAL_MALFORMED_ASN1_ENCOUNTERED);
```

Comment:
Add some tests to stress the failure cases.

They're all leaking `decoder` right now